### PR TITLE
chore(runtime): decouple consensus types

### PR DESCRIPTION
**Pull request by archbear**
part of larger proj to make modular runtime interface for diff types of consensus

## Summary by CodeRabbit

- **New Features**
  - Enhanced block handling and processing within the blockchain execution engine by introducing new payload and header types, improving robustness and efficiency.

- **Refactor**
  - Restructured and streamlined function signatures across multiple modules to include new types for better organization and maintainability.
  - Refactored various interfaces and types within the middleware package to accommodate new functionality and improve internal consistency.

- **Bug Fixes**
  - Addressed inconsistencies in method signatures and parameter types to ensure coherence across the system.

- **Documentation**
  - Updated function signatures and interface definitions to reflect recent changes, ensuring clarity and accuracy.

### CODE REVIEW

File: mod/runtime/pkg/middleware/abci.go [hunk 1]
```diff
@@ -105,11 +95,9 @@
...
 // prepareProposal is the internal handler for preparing proposals.
 func (h *ABCIMiddleware[
-	AvailabilityStoreT,
-	BeaconBlockT,
-	BeaconBlockBodyT,
-	BeaconStateT,
-	BlobSidecarsT,
+	AvailabilityStoreT, BeaconBlockT, BeaconBlockBodyT,
+	BeaconBlockHeaderT, BeaconStateT, BlobSidecarsT, DepositT,
+	Eth1DataT, ExecutionPayloadT, GenesisT,
```
[the remark this answers is not in the document]
Comment (ocnc2) [reply mod/runtime/pkg/middleware/abci.go:100]: i agree with u coderabbitai!!

[the remark this answers is not in the document]
Comment (ocnc2) [reply mod/runtime/pkg/middleware/abci.go:100]: u up?

[the remark this answers is not in the document]
Comment (ocnc2) [reply mod/runtime/pkg/middleware/abci.go:100]: hi

File: mod/beacon/blockchain/execution_engine.go [hunk 2]
```diff
@@ -115,14 +124,19 @@
...
+	DepositT,
+	ExecutionPayloadT,
+	ExecutionPayloadHeaderT,
+	GenesisT,
+	WithdrawalT,
 ]) sendNextFCUWithoutAttributes(
 	ctx context.Context,
 	blk BeaconBlockT,
-	lph *types.ExecutionPayloadHeader,
+	lph ExecutionPayloadHeader,
```
[the remark this answers is not in the document]
Comment (itsdevbear) [reply mod/beacon/blockchain/execution_engine.go:139]: @archbear

File: mod/beacon/blockchain/types.go [hunk 3]
```diff
@@ -48,38 +47,63 @@ type AvailabilityStore[BeaconBlockBodyT any, BlobSidecarsT any] interface {
...
+	GetSlot() math.Slot
+	// GetProposerIndex returns the proposer index of the beacon block.
+	GetProposerIndex() math.ValidatorIndex
+	// GetParentBlockRoot returns the parent block root of the beacon block.
+	GetParentBlockRoot() common.Root
+	// SetStateRoot sets the state root of the beacon block.
+	SetStateRoot(common.Root)
+	// GetStateRoot returns the state root of the beacon block.
+	GetStateRoot() common.Root
+
```
Comment (itsdevbear) [inline mod/beacon/blockchain/types.go:80]: remove space

File: mod/beacon/blockchain/types.go [hunk 4]
```diff
@@ -48,38 +47,63 @@ type AvailabilityStore[BeaconBlockBodyT any, BlobSidecarsT any] interface {
...
+] interface {
+	ssz.Marshallable
+	// NewWithVersion creates a new beacon block with the given parameters.
+	NewWithVersion(
+		slot math.Slot,
+		proposerIndex math.ValidatorIndex,
+		parentBlockRoot common.Root,
+		forkVersion uint32,
+	) (BeaconBlockT, error)
+
```
Comment (itsdevbear) [inline mod/beacon/blockchain/types.go:65]: remove space

File: mod/beacon/blockchain/types.go [hunk 5]
```diff
@@ -139,6 +167,44 @@ type EventFeed[EventT any] interface {
 	Send(event EventT) int
 }
 
+// ExecutionPayload is the interface for the execution payload.
+type ExecutionPayload[T any, WithdrawalT Withdrawal] interface {
+	Empty(uint32) T
```
Comment (itsdevbear) [inline mod/beacon/blockchain/types.go:172]: i could be wrong, but i believe this level of constraint on ExecutionPayloadT isn't needed in this service

File: mod/beacon/blockchain/types.go [hunk 6]
```diff
@@ -219,3 +316,11 @@ type TelemetrySink interface {
 	// identified by the provided keys.
 	MeasureSince(key string, start time.Time, args ...string)
 }
+
+// Withdrawal is the interface for the withdrawal.
+type Withdrawal interface {
+	GetIndex() math.U64
```
Comment (itsdevbear) [inline mod/beacon/blockchain/types.go:322]: godoc

File: mod/beacon/validator/types.go [hunk 7]
```diff
@@ -94,6 +99,11 @@ type BeaconBlockBody[
 	GetExecutionPayload() ExecutionPayloadT
 }
 
+// BeaconBlockHeader represents a beacon block header interface.
+type BeaconBlockHeader interface {
```
Comment (itsdevbear) [inline mod/beacon/validator/types.go:103]: not used, also can just use ssz.Marshallable everywhere

File: mod/beacon/validator/types.go [hunk 8]
```diff
@@ -171,6 +186,22 @@ type DepositStore[DepositT any] interface {
...
+	New(
+		depositRoot common.Root,
+		depositCount math.U64,
+		blockHash common.ExecutionHash,
+	) T
+}
+
+// ExecutionPayloadHeader represents the execution payload header interface.
+type ExecutionPayloadHeader interface {
+	GetTimestamp() math.U64
```
Comment (itsdevbear) [inline mod/beacon/validator/types.go:200]: godoc

File: mod/consensus-types/pkg/genesis/genesis.go [hunk 9]
```diff
@@ -57,6 +57,25 @@ type Genesis[
...
+	return g.Deposits
+}
+
+// GetExecutionPayloadHeader returns the execution payload header.
+func (g *Genesis[
+	DepositT, ExecutionPayloadHeaderT,
+]) GetExecutionPayloadHeader() ExecutionPayloadHeaderT {
+	return g.ExecutionPayloadHeader
+}
+
```
[the remark this answers is not in the document]
Comment (itsdevbear) [reply mod/consensus-types/pkg/genesis/genesis.go:78]: @archbear coderabbit and i agree with this safety check. 

alternatively create types.Deposits (plural) and a Validate() function.  (id do this)

[the remark this answers is not in the document]
Comment (itsdevbear) [reply mod/consensus-types/pkg/genesis/genesis.go:78]: we should also create a consistent pattern between all our types to ensure this. 

maybe all constructors check for validity? 

good to be consistent amongst all consensus types

**Review by itsdevbear — APPROVED** [review]

File: mod/runtime/pkg/middleware/abci.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ocnc) [reply mod/runtime/pkg/middleware/abci.go]: don't hit on the ai, they get really attached

**Review by itsdevbear — APPROVED** [review]

**Pull request merged**